### PR TITLE
Merge rules per ingress by the same host, pathType and backend service

### PR DESCRIPTION
**Pull request by oliviassss**
### Issue

https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2203
https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2678

### Description

Merged the rules per ingress if:

- they are routing to the same services, and
- they have the same host, and
- they are of the same pathType

#### Tests
- Added unit tests
- Created Ingress with multiple rules, verified that paths were merged only when they have the same host, pathType and backend service. Also verified that when the count of the condition values exceed 5, the merged paths will flow over to a new rule in order to fulfill the current quota limit for rules.

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/ingress/model_build_listener_rules.go [hunk 1]
```diff
@@ -3,40 +3,116 @@ package ingress
...
+				var paths []string
+				for _, path := range mergedPaths {
+					if path.Path != "" {
+						paths = append(paths, path.Path)
+					}
+				}
+				enhancedBackend, err := t.enhancedBackendBuilder.Build(ctx, ing.Ing, currBackend,
 					WithLoadBackendServices(true, t.backendServices),
 					WithLoadAuthConfig(true))
+				conditions, err := t.buildRuleConditions(ctx, host, paths, enhancedBackend, *currPathType)
```
Comment (johngmyers) [inline pkg/ingress/model_build_listener_rules.go]: Moving this causes any `err` returned from `enhancedBackendBuilder` to be ignored.

File: pkg/ingress/model_build_listener_rules.go [hunk 2]
```diff
@@ -119,20 +243,83 @@ func (t *defaultModelBuildTask) classifyIngressPathsByType(paths []networking.HT
 	return exactPaths, prefixPaths, implementationSpecificPaths, nil
 }
 
-func (t *defaultModelBuildTask) buildRuleConditions(ctx context.Context, rule networking.IngressRule,
-	path networking.HTTPIngressPath, backend EnhancedBackend) ([]elbv2model.RuleCondition, error) {
+// func (t *defaultModelBuildTask) buildRuleConditions(ctx context.Context, rule networking.IngressRule,
```
Comment (johngmyers) [inline pkg/ingress/model_build_listener_rules.go]: Edit this commented-out block from the commit

File: pkg/ingress/model_build_listener_rules.go [hunk 3]
```diff
@@ -77,6 +153,54 @@ func (t *defaultModelBuildTask) buildListenerRules(ctx context.Context, lsARN co
...
+			}
+		}
+	}
+	// iterate the mergeRefMap and append all paths in a group to the rule of the min ruleIdx
+	var mergedRules []networking.IngressRule
+	for _, mergedPaths := range mergeRefMap {
+		minIdx := pathToRuleMap[mergedPaths[0]]
+		for _, path := range mergedPaths {
+			currIdx := pathToRuleMap[path]
+			minIdx = algorithm.GetMin(minIdx, currIdx)
```
Comment (johngmyers) [inline pkg/ingress/model_build_listener_rules.go]: The Go way is to just inline this.

File: pkg/ingress/model_build_listener_rules.go [hunk 4]
```diff
@@ -149,16 +150,42 @@ func (t *defaultModelBuildTask) buildListenerRules(ctx context.Context, lsARN co
...
 	}
-
 	return nil
 }
 
-func (t *defaultModelBuildTask) mergeRulesPerIngress(rules []networking.IngressRule) ([]networking.IngressRule, error) {
-	// use a map to help merge rules by host, pathType and service
+// getRulesToMerge classifies rules into two categories - rules with unique host, rules with replicate hosts
+// only rules with replicate hosts need merge
+func (t *defaultModelBuildTask) getRulesToMerge(rules []networking.IngressRule) ([]networking.IngressRule, []networking.IngressRule, error) {
```
Comment (johngmyers) [inline pkg/ingress/model_build_listener_rules.go]: Perhaps name the return values?

File: pkg/ingress/model_build_listener_rules.go [hunk 5]
```diff
@@ -149,16 +150,42 @@ func (t *defaultModelBuildTask) buildListenerRules(ctx context.Context, lsARN co
...
+	var rulesWithUniqueHost []networking.IngressRule
+	hostToRulesMap := make(map[string][]networking.IngressRule)
+	for _, rule := range rules {
+		host := rule.Host
+		_, exists := hostToRulesMap[host]
+		if exists {
+			hostToRulesMap[host] = append(hostToRulesMap[host], rule)
+		} else {
+			hostToRulesMap[host] = []networking.IngressRule{rule}
+		}
```
Comment (johngmyers) [inline pkg/ingress/model_build_listener_rules.go:80]:
```
hostToRulesMap[host] = append(hostToRulesMap[host], rule)
```

File: pkg/ingress/model_build_listener_rules.go [hunk 6]
```diff
@@ -149,16 +150,42 @@ func (t *defaultModelBuildTask) buildListenerRules(ctx context.Context, lsARN co
...
-
 	return nil
 }
 
-func (t *defaultModelBuildTask) mergeRulesPerIngress(rules []networking.IngressRule) ([]networking.IngressRule, error) {
-	// use a map to help merge rules by host, pathType and service
+// getRulesToMerge classifies rules into two categories - rules with unique host, rules with replicate hosts
+// only rules with replicate hosts need merge
+func (t *defaultModelBuildTask) getRulesToMerge(rules []networking.IngressRule) ([]networking.IngressRule, []networking.IngressRule, error) {
+	var rulesWithReplicateHosts []networking.IngressRule
```
Comment (johngmyers) [inline pkg/ingress/model_build_listener_rules.go:70]: `rulesWithReplicatedHosts`

**Review by johngmyers — CHANGES_REQUESTED** [review]
Is there a big advantage to having a separate code path for unmerged listener rules? It seems like a lot of complexity and I worry about test coverage.

**Comment by omri-shilton** [comment]
@johngmyers This PR might solve a huge issue for us. also solves #2678 
Can we consider merging this?

**Pull request closed without merging**